### PR TITLE
fix(llmobs): replace trace processor with event listener

### PR DESCRIPTION
**Pull request by Kyle-Verhoog**
The LLMObs service formerly depended on the TraceProcessor interface in the tracer. This was problematic due to sharing a dependency with the public API. As such, users could configure a trace filter (under the hood is a trace processor) and overwrite the LLMObs TraceProcessor.

Instead, the tracer can emit span start and finish events which the LLMObs service listens to and acts on, as proposed here.

The gotcha is that the LLMObs service no longer has a way to drop traces when run in agentless mode, which only LLMObs supports. Instead, we encourage users to explicitly turn off APM which carries the benefit of clarity since this was implicit before.

[](https://datadoghq.atlassian.net/browse/MLOB-1925)

## Checklist
- [x] PR author has checked that all the criteria below are met
- The PR description includes an overview of the change
- The PR description articulates the motivation for the change
- The change includes tests OR the PR description describes a testing strategy
- The PR description notes risks associated with the change, if any
- Newly-added code is easy to change
- The change follows the [library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html)
- The change includes or references documentation updates if necessary
- Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist
- [x] Reviewer has checked that all the criteria below are met 
- Title is accurate
- All changes are related to the pull request's stated goal
- Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes
- Testing strategy adequately addresses listed risks
- Newly-added code is easy to change
- Release note makes sense to a user of the library
- If necessary, author has acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment
- Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)

### CODE REVIEW

**Comment by Kyle-Verhoog** [comment]
the existing tests for the trace processor are coupled to the implementation so i'm gonna open up a refactor on those first instead of trying to migrate them at the same time increasing the risk of a regression

**Comment by Kyle-Verhoog** [comment]
done in https://github.com/DataDog/dd-trace-py/pull/11784

File: tests/llmobs/test_propagation.py [hunk 1]
```diff
@@ -216,8 +216,6 @@ def test_activate_distributed_headers_propagate_correct_llmobs_parent_id_simple(
     env["DD_TRACE_ENABLED"] = "0"
     stdout, stderr, status, _ = run_python_code_in_subprocess(code=code, env=env)
     assert status == 0, (stdout, stderr)
-    assert stderr == b"", (stdout, stderr)
```
Comment (Yun-Kim) [inline tests/llmobs/test_propagation.py:219]: Note here there's a weird stderr output that keeps trying to submit traces to llmobs from the subprocess. I've attempted replacing the span writer with a dummy/mock, but this still happens which suggests there's some weird writer instantiation/assignment logic

Comment (Yun-Kim) [reply tests/llmobs/test_propagation.py:219]: Since this doesn't actually break any behavior and is an internal testing issue, I'm going to remove the breaking assertion for now and address this in a separate PR.

**Comment by Yun-Kim** [comment]
> Instead, we encourage users to explicitly turn off APM which carries the benefit of clarity since this was implicit before.

@Kyle-Verhoog should we add a release note specifying that `DD_TRACE_ENABLED` should be set to false for agentless users? Or would you rather we do this implicitly for agentless customers?

**Comment by Kyle-Verhoog** [comment]
yes let's add a release note for the change

File: ddtrace/llmobs/_llmobs.py [hunk 2]
```diff
@@ -160,11 +284,7 @@ def _stop_service(self) -> None:
         except ServiceStatusError:
             log.debug("Error stopping LLMObs writers")
 
-        try:
-            forksafe.unregister(self._child_after_fork)
-            self.tracer.shutdown()
-        except Exception:
-            log.warning("Failed to shutdown tracer", exc_info=True)
+        forksafe.unregister(self._child_after_fork)
```
Comment (lievan) [inline ddtrace/llmobs/_llmobs.py:287]: i noticed we removed `tracer.shutdown` here, is it because it's not necessary and the tracer will already be shutdown elsewhere?

File: tests/llmobs/test_llmobs_ragas_faithfulness_evaluator.py [hunk 3]
```diff
@@ -167,36 +170,34 @@ def test_ragas_faithfulness_submits_evaluation_on_span_with_custom_keys(ragas, L
 
 
 @pytest.mark.vcr_logs
-def test_ragas_faithfulness_emits_traces(ragas, LLMObs):
-    rf_evaluator = RagasFaithfulnessEvaluator(LLMObs)
+def test_ragas_faithfulness_emits_traces(ragas, llmobs, llmobs_events):
+    """Why are we asserting only 7 spans caught?"""
```
Comment (lievan) [inline tests/llmobs/test_llmobs_ragas_faithfulness_evaluator.py:174]: for this comment, i think we discussed this offline (autoinstrumented spans aren't emitted here)

File: ddtrace/llmobs/_utils.py [hunk 4]
```diff
@@ -135,9 +135,12 @@ def _get_ml_app(span: Span) -> str:
     ml_app = span._get_ctx_item(ML_APP)
     if ml_app:
         return ml_app
-    nearest_llmobs_ancestor = _get_nearest_llmobs_ancestor(span)
-    if nearest_llmobs_ancestor:
-        ml_app = nearest_llmobs_ancestor._get_ctx_item(ML_APP)
+    llmobs_parent = _get_nearest_llmobs_ancestor(span)
```
[the remark this answers is not in the document]
Comment (Kyle-Verhoog) [reply ddtrace/llmobs/_utils.py:138]: the issue is that the ml app was only being set after the span finished but we are hooking the span finish event now. So the ml app propagation logic happens _after_ we observe the span and we need to look all the way up the tree to find it.

I think I'd like to follow up here to store the ml app at a trace/contextual level but that was a bit out of scope here.

[the remark this answers is not in the document]
Comment (lievan) [reply ddtrace/llmobs/_utils.py:138]: Ah i see thx

[the remark this answers is not in the document]
Comment (Yun-Kim) [reply ddtrace/llmobs/_utils.py:138]: The issue is that now because we moved from trace processor (which went top-down from the root to its children and etc), and previously we were propagating ML App from a span to its children (which allowed us to just grab the immediate parent ml app). However now with the span start/finish hooks we now process the leaf spans first, meaning we need to now look for the nearest llmobs ancestor span that has an ml app specified in order to propagate ml app correctly. does that make sense?

File: tests/llmobs/test_llmobs_ragas_faithfulness_evaluator.py (hunk 3 above)
Comment (Yun-Kim) [reply tests/llmobs/test_llmobs_ragas_faithfulness_evaluator.py:174]: good catch, will remove this comment

File: ddtrace/llmobs/_llmobs.py (hunk 2 above)
Comment (Kyle-Verhoog) [reply ddtrace/llmobs/_llmobs.py:287]: hmm it's a good catch. I'm not sure why we needed to shutdown the tracer before... I don't see a need for us to do so, so if all our tests are passing still then I think it's fine.

Comment (Yun-Kim) [reply ddtrace/llmobs/_llmobs.py:287]: Yeah not a problem anymore, we only needed shutdown() to flush on disable in case of partial flushing and dependence on trace processors. Now with listener hooks we shouldn't have this issue.

File: ddtrace/_trace/tracer.py [hunk 5]
```diff
@@ -849,7 +850,7 @@ def _start_span(
             for p in chain(self._span_processors, SpanProcessor.__processors__, self._deferred_processors):
                 p.on_span_start(span)
         self._hooks.emit(self.__class__.start_span, span)
-
+        dispatch("trace.span_start", (span,))
```
Comment (mabdinur) [inline ddtrace/_trace/tracer.py:853]: I love this. Do you see a need for `tracer._hooks`? Are there are blockers to replacing this mechanism with core dispatch/on?

File: tests/llmobs/conftest.py [hunk 6]
```diff
@@ -191,16 +198,20 @@ def mock_ragas_dependencies_not_present():
 @pytest.fixture
 def ragas(mock_llmobs_span_writer, mock_llmobs_eval_metric_writer):
     with override_global_config(dict(_dd_api_key="<not-a-real-key>")):
-        import ragas
-
+        try:
+            import ragas
+        except ImportError:
```
Comment (mabdinur) [inline tests/llmobs/conftest.py:203]: Can we leave a removal note for when we drop support for 3.7. I don't want us to forget to clean up this code.

File: ddtrace/_trace/tracer.py (hunk 5 above)
Comment (mabdinur) [reply ddtrace/_trace/tracer.py:853]: Do we need this to run before or after the existing start/finish hooks? 

For example: With lazy sampling spans may not have a sampling decision until `TraceSamplingProccesor` is executed. Do LLM spans need a valid sampling decision? 

There might be other edge cases. It would be nice if the ordering of the hooks were well defined.

**Review by mabdinur — APPROVED** [review]
Left some questions but overall looks good to me

File: tests/llmobs/conftest.py (hunk 6 above)
Comment (Yun-Kim) [reply tests/llmobs/conftest.py:180]: Is this because we don't need to `try/except ImportError` in Python 3.8+?

File: ddtrace/_trace/tracer.py (hunk 5 above)
Comment (Yun-Kim) [reply ddtrace/_trace/tracer.py:853]: We shouldn't be blocked on any sampling decisions since we are decoupled from APM in that regard. So I think this should be fine where it is, since this is an independent data flow to LLM Observability.

Comment (Yun-Kim) [reply ddtrace/_trace/tracer.py:853]: > Do you see a need for tracer._hooks? Are there are blockers to replacing this mechanism with core dispatch/on?

We actually were previously using tracer._hooks, but this listener was simpler and made more sense for our use case. I don't see any blockers with replacing the _hooks with core dispatch unless there are specific hooks that you NEED to happen in a specific order (first/last, etc)

**Review by lievan — APPROVED** [review]
